tiledb/sm/filter/bitsort_filter.cc

```diff
@@ -0,0 +1,450 @@
+      return Status_FilterError(
+          "BitSortFilter::run_reverse: datatype does not have an appropriate "
+          "size");
+    }
+  }
+}
+
+template <typename AttrType>
+Status BitSortFilter::run_reverse(
+    BitSortFilterMetadataType &bitsort_metadata,
+    FilterBuffer* input_metadata,
+    FilterBuffer* input,
+    FilterBuffer* output_metadata,
+    FilterBuffer* output) const  {
+  // Get number of parts.
+  uint32_t num_parts;
+  RETURN_NOT_OK(input_metadata->read(&num_parts, sizeof(uint32_t)));
+
+  RETURN_NOT_OK(output->prepend_buffer(input->size()));
+  Buffer* output_buf = output->buffer_ptr(0);
+  assert(output_buf != nullptr);
+
+  // Get number of cells in the tile.
+  std::vector<Tile*> &dim_tiles = bitsort_metadata.first.get();
+  uint64_t cell_pos_size = dim_tiles[0]->cell_num();
+
+  // Determine the positions vector and rewrite the dimension tiles using these positions.
+  std::vector<uint64_t> cell_pos;
+  cell_pos.reserve(cell_pos_size);
+  run_reverse_dim_tiles(bitsort_metadata, cell_pos);
+
+  // Unsort the attribute data with the positions determined by the dimension global sorting.
+  for (uint32_t i = 0; i < num_parts; i++) {
+    uint32_t part_size;
+    RETURN_NOT_OK(input_metadata->read(&part_size, sizeof(uint32_t)));
+    ConstBuffer part(nullptr, 0);
+    RETURN_NOT_OK(input->get_const_buffer(part_size, &part));
+
+    RETURN_NOT_OK(unsort_part<AttrType>(cell_pos, &part, output_buf));
+
+    if (output_buf->owns_data()) {
+      output_buf->advance_size(part_size);
+    }
+    output_buf->advance_offset(part_size);
+    input->advance_offset(part_size);
+  }
+
+  // Output metadata is a view on the input metadata, skipping what was used
+  // by this filter.
+  auto md_offset = input_metadata->offset();
+  RETURN_NOT_OK(output_metadata->append_view(
+      input_metadata, md_offset, input_metadata->size() - md_offset));
+
+  return Status::Ok();
+}
+
+template <typename AttrType>
+Status BitSortFilter::unsort_part(
+    const std::vector<uint64_t> &cell_pos, ConstBuffer* input_buffer, Buffer* output_buffer) const {
+  // Read in data.
+  uint32_t buff_size = input_buffer->size();
+  assert(buff_size % sizeof(AttrType) == 0);
+  uint32_t num_elems_in_part = buff_size / sizeof(AttrType);
+
+  if (num_elems_in_part == 0) {
+    return Status::Ok();
+  }
+
+  const AttrType* input_array = static_cast<const AttrType*>(input_buffer->data());
+  AttrType* output_array = static_cast<AttrType*>(output_buffer->cur_data());
+
+  // Write in data in original order.
+  for (uint32_t i = 0; i < num_elems_in_part; ++i) {
+    output_array[i] = input_array[cell_pos[i]];
+  }
+
+  return Status::Ok();
+}
+
+Status BitSortFilter::run_reverse_dim_tiles(BitSortFilterMetadataType &bitsort_metadata, std::vector<uint64_t> &cell_pos) const {
+  std::vector<Tile*> &dim_tiles = bitsort_metadata.first.get();
+  auto cmp_fn = bitsort_metadata.second.get();
+
+  for (uint64_t i = 0; i < dim_tiles[0]->cell_num(); ++i) {
```

Comment:
Use iota as well to initialize cell_pos. Also run_reverse_dim_tiles should return cell_pos instead of taking it in as a parameter. We can also convert the function to not return status and throw exceptions.